src/test/java/com/cst438/controller/AssignmentControllerUnitTest.java

```diff
@@ -0,0 +1,157 @@
+                                    .contentType(MediaType.APPLICATION_JSON)
+                                    .content(asJsonString(student2)))
+                    .andReturn()
+                    .getResponse();
+
+            // Create user for enterFinalGradesSuccess
+            UserDTO student3 = new UserDTO(0, "Student Three", "student3@example.com", "STUDENT");
+            mvc.perform(
+                            MockMvcRequestBuilders
+                                    .post("/users")
+                                    .contentType(MediaType.APPLICATION_JSON)
+                                    .content(asJsonString(student3)))
+                    .andReturn()
+                    .getResponse();
+
+            setUpIsDone = true;
+        }
+    }
+
+    @Test
+    public void gradeAssignmentSuccess() throws Exception {
+        MockHttpServletResponse response = mvc.perform(
+                        MockMvcRequestBuilders
+                                .get("/assignments/1/grades")
+                                .contentType(MediaType.APPLICATION_JSON))
+                .andReturn()
+                .getResponse();
+
+        assertEquals(200, response.getStatus(), "Expected status 200 but got " + response.getStatus());
+
+        GradeDTO[] grades = fromJsonString(response.getContentAsString(), GradeDTO[].class);
+        GradeDTO[] updatedGrades = new GradeDTO[grades.length];
+        for (int i = 0; i < grades.length; i++) {
+            GradeDTO grade = grades[i];
+            updatedGrades[i] = new GradeDTO(grade.gradeId(), grade.studentName(), grade.studentEmail(), grade.assignmentTitle(), grade.courseId(), grade.sectionId(), grade.score() + 5); // Example modification
+        }
+
+        response = mvc.perform(
+                        MockMvcRequestBuilders
+                                .put("/grades")
+                                .content(asJsonString(updatedGrades))
+                                .contentType(MediaType.APPLICATION_JSON))
+                .andReturn()
+                .getResponse();
+
+        assertEquals(200, response.getStatus(), "Expected status 200 but got " + response.getStatus());
+    }
+
+    @Test
+    public void gradeAssignmentInvalidAssignmentId() throws Exception {
+        MockHttpServletResponse response;
+
+        List<GradeDTO> grades = List.of(
+                new GradeDTO(9999, "Student One", "student1@example.com", "Assignment One", "CST438", 1, 95),
+                new GradeDTO(9999, "Student Two", "student2@example.com", "Assignment One", "CST438", 1, 85)
+        );
+
+        response = mvc.perform(
+                        MockMvcRequestBuilders
+                                .put("/grades")
+                                .content(asJsonString(grades))
+                                .contentType(MediaType.APPLICATION_JSON))
+                .andReturn()
+                .getResponse();
+
+        assertEquals(404, response.getStatus());
```

Comment:
I don't think this test is quite right. The rubric says we are supposed to invoke the REST api to GET assignment grades using URL /assignments/{assignment_id}/grades with an invalid assignment id.

Reply:
However, the Google doc does say "instructor attempts to grade an assignment but the assignment id is invalid" which yours is in line with. I would maybe add another test rather than removing this one just so our bases are covered.

Reply:
I don't see any changes made here. You still need to change this one.

Reply:
I like this. It covers multiple scenarios well.